### PR TITLE
APIGW: use "cue" system for APIGW too

### PR DESCRIPTION
**Pull request by justinmk3**
- "sam local start-api" outputs similar "Starting debugger" messages as  "sam local invoke", so we can reuse the `SamCliLocalInvokeInvocation`  logic related to that.
- Fix the `runDebugger` promise so that it returns the  `childProcess.start()` promise rather than an async promise. Else
  `.then(() => { timeExpired = false }` does not work (awaiting the  `runDebugger` promise would not await the inner promise).

## License

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: src/shared/sam/cli/samCliLocalInvoke.ts [hunk 1]
```diff
@@ -76,10 +75,9 @@ export class DefaultSamLocalInvokeCommand implements SamLocalInvokeCommand {
                     // If we have a timeout (as we do on debug) refresh the timeout as we receive text
                     params.timeout?.refresh()
                     this.logger.verbose(`SAM: pid ${childProcess.pid()}: stderr: ${text}`)
-                    if (checkForDebuggerAttachCue) {
+                    if (checkForCues) {
                         // Look for messages like "Waiting for debugger to attach" before returning back to caller
                         if (this.debuggerAttachCues.some(cue => text.includes(cue))) {
-                            checkForDebuggerAttachCue = false
```
Comment (justinmk3) [inline src/shared/sam/cli/samCliLocalInvoke.ts:82]: don't need to assign this, we resolve() below

File: src/shared/sam/cli/samCliLocalInvoke.ts [hunk 2]
```diff
@@ -106,14 +104,15 @@ export class DefaultSamLocalInvokeCommand implements SamLocalInvokeCommand {
...
+                        'Error running command "{0}": {1}',
+                        samCommandName,
+                        error
+                    )
                     reject(error)
                 },
             })
-
-            if (!params.isDebug || this.debuggerAttachCues.length === 0) {
-                resolve()
```
Comment (justinmk3) [inline src/shared/sam/cli/samCliLocalInvoke.ts:115]: not sure what the purpose of this was. If the caller of `invoke()` does not want to wait they can just omit `await`.

File: src/shared/sam/cli/samCliLocalInvoke.ts [hunk 3]
```diff
@@ -122,7 +121,11 @@ export class DefaultSamLocalInvokeCommand implements SamLocalInvokeCommand {
 
         await Promise.race(awaitedPromises).catch(async () => {
             if (timeExpired) {
-                this.channelLogger.error('AWS.samcli.timeout', 'Timeout while waiting for command: "{0}"', samCommandName)
+                this.channelLogger.error(
+                    'AWS.samcli.timeout',
+                    'Timeout while waiting for command: "{0}"',
+                    samCommandName
```
Comment (justinmk3) [inline src/shared/sam/cli/samCliLocalInvoke.ts:127]: don't know why prettier reformatted this

File: src/shared/sam/localLambdaRunner.ts [hunk 4]
```diff
@@ -319,9 +319,9 @@ export async function invokeLambdaFunction(
     if (!config.noDebug) {
         if (config.invokeTarget.target === 'api') {
             const payload = JSON.parse(await readFile(config.eventPayloadFile, { encoding: 'utf-8' }))
-            await requestLocalApi(ctx, config.api!, config.apiPort!, payload).finally(() => {
-                stopApi(config)
-            })
+            await requestLocalApi(ctx, config.api!, config.apiPort!, payload)
+            // Wait for cue messages ("Starting debugger" etc.).
+            await samStartApi
```
Comment (justinmk3) [inline src/shared/sam/localLambdaRunner.ts:324]: this improves reliability of APIGW debugging, especially for cases where downloading an image takes a long time.

File: src/shared/sam/localLambdaRunner.ts [hunk 5]
```diff
@@ -66,11 +66,11 @@ function makeResourceName(config: SamLaunchRequestArgs): string {
 
 const SAM_LOCAL_PORT_CHECK_RETRY_INTERVAL_MILLIS: number = 125
 const SAM_LOCAL_PORT_CHECK_RETRY_TIMEOUT_MILLIS_DEFAULT: number = 30000
-const MAX_DEBUGGER_RETRIES_DEFAULT: number = 30
-const ATTACH_DEBUGGER_RETRY_DELAY_MILLIS: number = 200
+const MAX_DEBUGGER_RETRIES_DEFAULT: number = 4
```
Comment (justinmk3) [inline src/shared/sam/localLambdaRunner.ts:69]: These retries never happened until it was fixed in https://github.com/aws/aws-toolkit-vscode/pull/1289 . But if it does retry, it is not a great UX (modal dialog). So, don't do 30 of them. Should rarely be needed.

File: src/shared/sam/localLambdaRunner.ts [hunk 6]
```diff
@@ -363,20 +363,20 @@ export async function invokeLambdaFunction(
...
     }
 
     try {
-        getLogger().verbose('SAM: sending SIGHUP to sam process: pid %d', samStartApi.pid())
-        samStartApi.stop(true, 'SIGHUP')
+        getLogger().verbose('SAM: sending SIGHUP to sam process: pid %d', proc.pid())
+        proc.stop(true, 'SIGHUP')
     } catch (e) {
-        getLogger().warn('SAM: failed to stop sam process: pid %d: %O', samStartApi.pid(), e as Error)
+        getLogger().warn('SAM: failed to stop sam process: pid %d: %O', proc.pid(), e as Error)
```
Comment (bryceitoc9) [inline src/shared/sam/localLambdaRunner.ts:377]: Should we consider this a failure? Should we attempt a `SIGKILL`?

**Review by bryceitoc9 — APPROVED** [review]

File: src/shared/sam/localLambdaRunner.ts (hunk 6 above)
Comment (justinmk3) [reply src/shared/sam/localLambdaRunner.ts:377]: `proc.stop(true, ...)` does SIGKILL implicitly if the process does not exit after some time. This try-catch is just defensive :)

**Pull request merged**